Method Single::reduce() added

Smoren opened the pull request:
New method is fully covered by tests.
It seemed to me that this method would be suitable for your beautiful repository.
I hope you will agree with me.
Thanks for this great package!

Smoren committed:
Method Single::reduce() added

markrogoyski commented:
Hi @Smoren,

Thank you for your interest in IterTools and your kind words.

Up until now, all of the methods are providing functionality to iterate, so the call to a function returns something that itself can be iterated.

Reduce is interesting, because the result is a single value, so the call to the function doesn't return an iterator, but the result of iteration and reduction.

I think it makes sense to create a new namespace class for this, since it is different than current iteration categories. How about a class `IterTools\Reduce` and the method that you are submitting can be `Reduce::toValue(...)`.

There are other reductions I am considering, such as `Reduce::toLength`, `Reduce::toMin/Max`, `Reduce::isSorted`, `Reduce::exactlyN` etc., so I think having it's own namespace class is the right place to put this functionality since it is different from other namespaces that provide iteration rather than single values.

Some tips for the unit tests: Structure the tests in the Given/When/Then format using those comments. Take a look at how other tests are formatted. Also, when asserting an expected value, have the result be precomputed. Do not use code in a unit test. We write unit tests to validate _our_ code. We don't want to use code to validate code--we want to use specific answers to validate code. Again, take a look at other tests in this library to see how results are precomputed by hand and fed into assertions as expected answers. Also, try to think of edge/boundary cases: empty list, list of single value, carry not provided, (should carry be required?).

Thanks again for your pull request and introducing some new functionality.
Mark

Smoren committed:
1. Method `Single::reduce()` removed
2. Class `Reduce` created with methods:
 - toValue
 - toMin
 - toMax
 - toCount
 - toSum
 - toProduct
 - toAverage
3. Tests added for all new methods (coverage 100%).
4. README.md updated with description of new methods.
5. Bug fixed in `ArrayIteratorFixture` at line 50. See comment there.

Smoren commented:
Thank you for your feedback!

Hope it's done right now.

1. Method `Single::reduce()` removed
2. Class `Reduce` created with methods:
 - toValue
 - toMin
 - toMax
 - toCount
 - toSum
 - toProduct
 - toAverage
3. Tests added for all new methods (coverage 100%).
4. README.md updated with description of new methods.
5. Bug fixed in `ArrayIteratorFixture` at line 50. See comment there.

markrogoyski commented:
I think you will need to add a new Reduce test directory to the PHPUnit XML configuration for tests and code coverage to run on the new code. See: https://github.com/markrogoyski/itertools-php/blob/main/tests/phpunit.xml

Smoren committed:
Reduce directory added to phpunit.xml

Smoren commented:
Done. I'm sorry, I'm used to the codeception, so I didn't figure it out right away.

markrogoyski commented on src/Reduce.php:
```diff
@@ -0,0 +1,128 @@
+     *
+     * @return T
+     */
+    public static function toValue(iterable $data, callable $reducer, $carry = null)
```
The third parameter here, `$carry` is exposing an implementation detail. I think a better abstraction for this from the client point of view is as the "initial value." So, what do you think about renaming this to `$initialValue` and then initialize a `$carry` variable prior to the foreach loop?

Smoren replied:
Totally agree. Done.

Smoren committed:
3-rd parameter of Reduce::toValue() renamed:  =>

Smoren committed:
Method Reduce::toCount() optimized for arrays and Countable objects.

Smoren commented:
Hi!

I've optimized method Reduce::toCount() for arrays and Countable instances.
Now this method uses count() function for them, so complexity reduced from O(N) to O(1) if we get array or Countable as input.

Smoren committed:
Method Single::eachPair() added.

Smoren commented:
Also I've added method Single::eachPair() for iterating pair of elements in collection.
I need this method to implement Reduce::isSorted().

Smoren committed:
Methods Reduce::isSortedDirectly(), Reduce::isSortedReversely() added

Smoren committed:
Method Single::filterUnique() added.

Smoren committed:
Merge pull request #1 from Smoren/unique

Method Single::filterUnique() added.

Smoren committed:
README fix for Single::filterUnique()

Smoren commented:
A few more methods added:

- Reduce::isSortedDirectly(),
- Reduce::isSortedReversely(),
- Single::filterUnique().

I hope you think I'm still on the right way :)

Smoren committed:
Single::filterUnique(): mapping fix

Smoren committed:
Single::filterUnique(): more mapping fixes for non-strict comparisons of scalar types.
More tests added. Now looking OK.

Smoren committed:
garbage removed

Smoren commented:
@markrogoyski

Hi!

I'm sorry, I've made a lot of commits, so I decided to make a short report so as not to waste a lot of your time.

There will be no more commits to this branch until your comments are received. Ideas for new methods will be worked out separately.

1. Here are all the methods that have been added:

- `Reduce::toValue()`;
- `Reduce::toMin()`;
- `Reduce::toMax()`;
- `Reduce::toCount()`;
- `Reduce::toSum()`;
- `Reduce::toProduct()`;
- `Reduce::toAverage()`;
- `Reduce::isSortedDirectly()`;
- `Reduce::isSortedReversely()`;
- `Single::filterUnique()`;
- `Single::eachPair()`.

2. All methods have been documented (in PHPdoc and README.md) and tested similarly to yours.

3. During testing, it was necessary to introduce a new fixture — `CountableIteratorAggregateFixture`, because the `Reduce::toCount()` method works in a special way with `Countable` entities (in order to optimize).

4. A bug was found and fixed in `ArrayIteratorFixture::valid()`. 

You used the `isset()` function there, which led to the premature termination of the iteration process, when the iterable array contained `null` (if `$a[0] === null` then `isset($a[0]) === false`).

I replaced `isset()` with `array_key_exists()` to fix this bug.

There is a comment in that method (file `ArrayIteratorFixture.php`, line 50) that needs to be removed before accepting this pull request.

5. A couple of questions:
- What do you think about renaming: `Single::eachPair()` to `Single::pairs()`?
- What do you think about renaming: `Single::toCount()` to `Single::toLength()`?

I will be very happy about feedback. I hope that my pull request will be accepted and the new methods will be useful to the community.

I am ready for any comments and I will make any necessary changes to the code according to your wishes.

Thank you for your time!

markrogoyski commented:
Hi @Smoren,

Thank you for implementing additional methods to fill out the new `Reduce` iter tools, and thanks for the summary of changes. I will review the code and leave any feedback I have.

One high-level comment. First off, thank you for adding unit tests. For tests that reduce to a boolean, such as the `isSorted` methods, I think the tests will be easier to follow and allow for simpler data provider methods if the "true" and "false" test cases are separated into separate test methods with their own data providers. Doing this, the data provider is only the values, and the test can simply `assertTrue` or `assertFalse`. For example, separate `IsSortedDirectlyTest` `testArray` into two test case methods: `testArrayTrue` and `testArrayFalse`. No expected parameter would be required as the expectation is always the same and tested directly with the `assertTrue`/`assertFalse` methods.

Thanks again for adding additional functionality and following the coding standards.
Mark

markrogoyski commented on src/Reduce.php:
```diff
@@ -0,0 +1,178 @@
+     */
+    public static function toCount(iterable $data): int
+    {
+        if (is_array($data) || $data instanceof \Countable) {
```
FYI: I think this can be replaced with a single call to [`is_countable`](https://www.php.net/manual/en/function.is-countable.php).

Smoren replied:
Done.

markrogoyski commented on src/Single.php:
```diff
@@ -171,4 +229,32 @@ public static function takeWhile(iterable $data, callable $predicate): \Generato
+     *
+     * @return \Generator<array{T, T}>
+     */
+    public static function eachPair(iterable $data): \Generator
```
I think the more-commonly recognized name for this functionality is [pairwise](https://en.wikipedia.org/wiki/Pairwise). How about renaming to that?

Smoren replied:
Done.

markrogoyski commented on src/Reduce.php:
```diff
@@ -0,0 +1,178 @@
+     *
+     * @return bool
+     */
+    public static function isSortedDirectly(iterable $data): bool
```
Suggestion: Let's just call this `isSorted`, and the reversed one `isReversed`.

Smoren replied:
Done.

markrogoyski commented on src/Single.php:
```diff
@@ -130,6 +130,64 @@ public static function filterTrue(iterable $data, callable $predicate = null): \
+     *
+     * @return \Generator<T>
+     */
+    public static function filterUnique(iterable $data, bool $strict = true): \Generator
```
I like this functionality and think it is a good addition. However, this is a bit complicated due to PHP's loose vs strict comparisons. I'd like to spend more time thinking about this and what the expected behavior should be. How about removing `filterUnique` from this pull request, and submit it again as a separate pull request with just that functionality? This way it will not block the rest of this pull request. Thanks.

Smoren replied:
Here is a new pull request for this method: https://github.com/markrogoyski/itertools-php/pull/4

markrogoyski commented on tests/Reduce/ToAverageTest.php:
```diff
@@ -0,0 +1,215 @@
+            [   [null, false],          0      ],
+            [   [true, false],          0.5    ],
+            [   [false, true],          0.5    ],
+            [   [0, null, false],       0      ],
+            [   [1, null, false],       0.3333 ],
+            [   [1, null, true],        0.6667 ],
```
Seeing this test case, I realize that the existing `Math` namespace functions like `runningAverage` will exhibit the same behavior. No need to make a change here now, but I think after this is merged I may go back to the Math functions and change them to throw an exception if a non-numeric value is provided, and do the same for these here. Doing math on booleans and nulls like they were legitimate numbers doesn't seem like the right thing to do. I will make that change later. Keep your code as it is as it is consistent with existing functionality for now.

Smoren replied:
Understood, accepted.

markrogoyski commented on tests/Reduce/ToAverageTest.php:
```diff
@@ -0,0 +1,215 @@
+        // When
+        $result = Reduce::toAverage($data);
+
+        $this->roundIfFloat($expected, $result);
+
+        // Then
+        $this->assertSame($expected, $result);
```
Suggestion: You can make use of the [assertEqualsWithDelta](https://phpunit.readthedocs.io/en/9.5/assertions.html#assertequalswithdelta) test method to simplify tests that have floating point values and not need to perform any arithmetic in between the method under test and the assertion, and thus, eliminating any extra code that might hide or introduce a bug in the code or tests.

Smoren replied:
Done.

Smoren committed:
- `Single::eachPair()` renamed to `Single::pairwise()`;
- `Reduce::isSortedDirectly()` renamed to `Reduce::isSorted()`;
- `Reduce::isSortedReversely()` renamed to `Reduce::isReversed()`;
- `Reduce::toCount()`: used function `is_countable()`.

Smoren committed:
Tests refactored with `assertTrue()` and `assertFalse()` for `Reduce::isSorted()` and `Reduce::isReverted()`.

Smoren committed:
`Single::filterUnique()` removed.

Smoren committed:
Merge pull request #2 from Smoren/dev

Refactoring by review

Smoren committed:
Tests refactored with `assertEqualsWithDelta()`: `ToAverageTest`, `ToSumTest`, `ToProductTest`.

Smoren committed:
Merge pull request #3 from Smoren/dev

Tests refactored with `assertEqualsWithDelta()`

Smoren commented:
@markrogoyski

Hi! Thank you for your feedback!

I've made all the suggested changes. Here is the summary:

 - `Single::eachPair()` renamed to `Single::pairwise()`;
 - `Reduce::isSortedDirectly()` renamed to `Reduce::isSorted()`;
 - `Reduce::isSortedReversely()` renamed to `Reduce::isReversed()`;
 - `Reduce::toCount()`: used function `is_countable()`;
 - Tests refactored with `assertTrue()` and `assertFalse()` for `Reduce::isSorted()` and `Reduce::isReverted()`;
 - `Single::filterUnique()` removed;
 - Tests refactored with `assertEqualsWithDelta()`: `ToAverageTest`, `ToSumTest`, `ToProductTest`.

markrogoyski commented:
Hi @Smoren,

Thanks again for the pull request. I've merged it and will include it in the next release of IterTools.

Thanks again!
Mark

Smoren commented:
@markrogoyski thank you very much!
I will continue thinking about new methods :)
